From 8c74af9e8556aea71eca995527b85f8f7bbbce4a Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:14:39 -0400
Subject: [net] udp: Verify UDP checksum before handoff to encap

Message-id: <1409652920-431-17-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90144
O-Subject: [PATCH 16/57] [RHEL7.1 net] net: Verify UDP checksum before handoff to encap
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit 0a80966b1043c3e2dc684140f155a3fded308660
Author: Tom Herbert <therbert@google.com>
Date:   Wed May 7 16:52:39 2014 -0700

    net: Verify UDP checksum before handoff to encap

    Moving validation of UDP checksum to be done in UDP not encap layer.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/udp.c b/net/ipv4/udp.c
index 5f54da2..770dc8d 100644
--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -1473,6 +1473,10 @@ int udp_queue_rcv_skb(struct sock *sk, struct sk_buff *skb)
   if (skb->len > sizeof(struct udphdr) && encap_rcv != NULL) {
    int ret;
 
+   /* Verify checksum before giving to encap */
+   if (udp_lib_checksum_complete(skb))
+    goto csum_error;
+
    ret = encap_rcv(sk, skb);
    if (ret <= 0) {
     UDP_INC_STATS_BH(sock_net(sk),
diff --git a/net/ipv6/udp.c b/net/ipv6/udp.c
index 89b5bea..eab6ebd 100644
--- a/net/ipv6/udp.c
+++ b/net/ipv6/udp.c
@@ -604,6 +604,10 @@ int udpv6_queue_rcv_skb(struct sock *sk, struct sk_buff *skb)
   if (skb->len > sizeof(struct udphdr) && encap_rcv != NULL) {
    int ret;
 
+   /* Verify checksum before giving to encap */
+   if (udp_lib_checksum_complete(skb))
+    goto csum_error;
+
    ret = encap_rcv(sk, skb);
    if (ret <= 0) {
     UDP_INC_STATS_BH(sock_net(sk),
-- 
1.7.1